File: 2019/hitcon/suicune/solve.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2019 hzshang <hzshang15@gmail.com>
import sys
MAX_LONG = (1<<64)-1
class Rand:
    def __init__(self,append,value):
        self.append = append
        self.value = (value * 0x5851F42D4C957F2D + 0x5851F42D4C957F2E) & MAX_LONG
        self.wtf = 0x9e
    def get_rand(self):
        r = self.value
        self.value = (self.value * 0x5851F42D4C957F2D + self.append) & MAX_LONG
        tmp = ((r ^ (r >> 18))>>27)&((1<<32)-1)
        offset = r >> 59
        return (tmp>>offset)|((tmp&((1<<offset)-1))<<(32-offset))

def encrypt(flag,key):
    rand = Rand(1,key)
    def shift(t):
        for i in range(len(t)-1,0,-1):
            v = rand.get_rand()%(i+1)
            t[v],t[i] = t[i],t[v]
        return t
    def get_table(length):
        table = [x for x in range(0x100)]
        table = shift(table)
        rand.get_rand()
        rand.get_rand()
        tmp = table[:length]
        tmp.sort()
        return tmp[::-1]
    s1 = flag
    for i in range(0x10):
        s2 = get_table(len(s1))
        data = "".join([chr(ord(x)^y) for x,y in zip(s1,s2)])
        s1 = data[::-1]
    return s1

def decrypt(enc,key):
    rand = Rand(1,key)
    def shift(t):
        for i in range(len(t)-1,0,-1):
            v = rand.get_rand()%(i+1)
            t[v],t[i] = t[i],t[v]
        return t
    def get_table(length):
        table = [x for x in range(0x100)]
        table = shift(table)
        rand.get_rand()
        rand.get_rand()
        tmp = table[:length]
        tmp.sort()
        return tmp[::-1]
    pool = []
    for i in range(0x10):
        pool.append(get_table(len(enc)))
    s1 = enc
    for i in range(0x10):
        s2 = pool[-i]
        data = "".join(chr(ord(x)^y) for x,y in zip(s1,s2))
        s1 = data[::-1]
    return s1

enc = "04dd5a70faea88b76e4733d0fa346b086e2c0efd7d2815e3b6ca118ab945719970642b2929b18a71b28d87855796e344d8".decode("hex")
enc = 
#enc =  encrypt("hitcon{1123213}",1234)
enc = encrypt("b"+"c"*47+"a",1234)
print enc.encode("hex")

File: 2019/hitcon/suicune/solve.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <stdint.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>


#include "global.h"

size_t hexs2bin(const char *hex, unsigned char **out);
typedef struct Rand_{
    uint64_t append;
    uint64_t value;
} Rand;
int comp(const void * elem1, const void * elem2) 
{
    uint8_t f = *(uint8_t*)elem1;
    uint8_t s = *(uint8_t*)elem2;
    if (f > s) return  1;
    if (f < s) return -1;
    return 0;
}
uint32_t randval(Rand *r){
    uint64_t val = r->value;
    r->value = r->value*0x5851F42D4C957F2D + r->append;
    uint32_t tmp = (val^(val>>18))>>27;
    uint32_t offset = val >> 59;
    return (tmp>>offset) | (((tmp&(1<<offset)-1))<<(32-offset));
}
void init_rand(Rand* r,uint16_t seed){
    r->append = 1;
    r->value = ((uint32_t)seed) * 0x5851F42D4C957F2D + 0x5851F42D4C957F2E;
}
Rand seed;
uint8_t pool[0x10][0x100];
uint64_t rand_table(uint8_t *table,int length){
    for(int i=length-1;i>=1;i--){
        int v = randval(&seed)%(i+1);
        uint8_t tmp = table[v];
        table[v] = table[i];
        table[i] = tmp;
    }
    uint64_t a = randval(&seed);
    uint64_t b = randval(&seed);
    uint64_t counts = (b<<32)|a;
    return counts;//wtf!!!
}
void reverse_array(uint8_t *array,int length);

void decrypt(uint8_t* enc,uint8_t* pt,int length,uint16_t key){
    init_rand(&seed,key);
    for(int i=0;i<0x10;i++){
        memcpy(pool[i],origin,sizeof(origin));
        uint64_t max_cost = rand_table(pool[i],0x100);
        uint8_t wtf[0x100];
        
//        memcpy(wtf,pool[i],0x100);
        calc_array_status(pool[i],length,max_cost);
        reverse_array(pool[i],length);
//        qsort(pool[i],length,1,comp);
//        asm("int $3");
    }
    uint8_t *tmp = malloc(length);
    memcpy(pt,enc,length);
    for(int i=0;i<0x10;i++){
        memcpy(tmp,pt,length);
        for(int j=0;j<length;j++){
            pt[length-1-j] = tmp[j]^pool[0xf-i][j];
        }
    }
    free(tmp);
}
int printable(uint8_t* data,int length){
    for(int i=0;i<length;i++){
        if(!isprint(data[i]))
            return 0;
    }
    return 1;
}


int main(int argc,char* argv[]){
    uint8_t buffer[0x100];
    memset(buffer,0,0x100);
    uint8_t *enc = "04dd5a70faea88b76e4733d0fa346b086e2c0efd7d2815e3b6ca118ab945719970642b2929b18a71b28d87855796e344d8";
    uint8_t* ct;
    int length = strlen(enc)/2;
    hexs2bin(enc,&ct);
    for(uint32_t key=0;key<0x10000;key++){
        memset(buffer,0,0x100);
        decrypt(ct,buffer,length,key);
        printf("%s\n",buffer);
        if(key%0x100 == 0){
            fprintf(stderr,"key 0x%04x\n",key);
        }
    }
}


